xbstrap/base.py

```diff
@@ -530,8 +530,37 @@ def cargo_home(self):
 class RequirementsMixin:
 	@property
 	def source_dependencies(self):
-		if 'sources_required' in self._this_yml:
-			yield from self._this_yml['sources_required']
+		sources_seen = set()
+		sources_stack = []
+
+		def visit(source):
+			if source in sources_seen:
+				return
+			sources_seen.add(source)
+			sources_stack.append(source)
+
+		def visit_yml(yml):
+			if isinstance(yml, dict):
+				visit(yml['name'])
+
+				if yml.get('recursive', False):
+					for source in self._cfg.get_source(yml['name'])._this_yml.get('sources_required', []):
+						visit(yml['name'])
+			else:
+				assert isinstance(yml, str)
+				visit(yml)
+
+		# Recursively visit all sources
+		for yml in self._this_yml.get('sources_required', []):
+			visit_yml(yml)
+
+		while sources_stack:
+			source = sources_stack.pop()
+			assert isinstance(source, str)
+			yield source
+
+			for yml in self._cfg.get_source(source)._this_yml.get('sources_required', []):
+				visit_yml(yml)
```

Comment:
To get the same behavior as for `tools:`, `recursive:` needs to be processed here, and not in visit_yml.

That would mean that in the bootstrap PR, the *dependencies* of `rustc-patched-libs` need to be annotated with `recursive: true` (but `rustc-patched-libs` itself does not need to be annotated when it is used a a dependency).